app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala

```diff
@@ -68,12 +68,25 @@ class SubscriptionService @Inject() (
         address.countryCode))
   }
 
-  def subscribeAgentToMtd(subscriptionRequest: SubscriptionRequest, authIds: AuthIds)(implicit hc: HeaderCarrier, ec: ExecutionContext, request: Request[Any]): Future[Option[Arn]] = {
-
+  def createSubscription(subscriptionRequest: SubscriptionRequest, authIds: AuthIds)(implicit hc: HeaderCarrier, ec: ExecutionContext, request: Request[Any]): Future[Option[Arn]] = {
     desConnector.getRegistration(subscriptionRequest.utr) flatMap {
-      case Some(DesRegistrationResponse(isAnAsAgent, _, _, maybeArn, BusinessAddress(_, _, _, _, Some(desPostcode), _), _)) if postcodesMatch(desPostcode, subscriptionRequest.knownFacts.postcode) => {
-        subscribe(subscriptionRequest, authIds, isAnAsAgent, maybeArn)
-      }
+      case Some(DesRegistrationResponse(isAnAsAgent, _, _, maybeArn, BusinessAddress(_, _, _, _, Some(desPostcode), _), _)) if postcodesMatch(desPostcode, subscriptionRequest.knownFacts.postcode) =>
+        for {
+          arn <- maybeArn match {
+            case Some(arn) if isAnAsAgent => Future.successful(arn)
+            case _ => desConnector.subscribeToAgentServices(subscriptionRequest.utr, desRequest(subscriptionRequest))
+          }
+          _ <- addKnownFactsAndEnrol(arn, subscriptionRequest, authIds)
+          _ <- {
+            subscriptionRequest.amlsDetails match {
```

Comment:
I think these lines 81-84 need to be executed at the point just before the call to the desConnector.subscribeToAgentServices on line 77.

Reply:
And we also need to call `agentAssuranceConnector.updateAmls` after the call to `desConnector.subscribeToAgentServices` (and before the `addKnownFactsAndEnrol`)

Reply:
but `Arn` is not available before `desConnector.subscribeToAgentServices`

Reply:
ah..sorry.. hold on